src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(kInvalidEndpointId != mapping.endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    FabricData fabric(fabric_index);
+
+    if (CHIP_NO_ERROR != fabric.Load(mStorage) || kInvalidEndpointId == fabric.first_endpoint)
+    {
+        // New fabric, or new endpoint
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group).Save(mStorage);
+        fabric.first_endpoint = mapping.endpoint;
+        fabric.endpoint_count = 1;
+        return fabric.Save(mStorage);
+    }
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            // Endpoint info not found
+            break;
+        }
+        if (endpoint.id == mapping.endpoint)
+        {
+            // Target endpoint info found
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    if (endpoint.id != mapping.endpoint)
+    {
+        // Endpoint not found, create new, and insert first
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group, fabric.first_endpoint).Save(mStorage);
+        fabric.first_endpoint = mapping.endpoint;
+        fabric.endpoint_count++;
+        return fabric.Save(mStorage);
+    }
+
+    // Endpoint info found, loop endpoints
+
+    GroupData group(fabric_index, endpoint.id, endpoint.first_group);
+
+    do
+    {
+        if (CHIP_NO_ERROR != group.Load(mStorage))
+        {
+            break;
+        }
+        if (group.group == mapping.group)
+        {
+            // Duplicated group
+            return CHIP_NO_ERROR;
+        }
+        group.group = group.next;
+    } while (kUndefinedGroupId != group.group);
+
+    // Insert group first, update endpoint
+    GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name, endpoint.first_group).Save(mStorage);
+    endpoint.first_group = mapping.group;
+    return endpoint.Save(mStorage);
+}
+
+CHIP_ERROR GroupDataProviderImpl::RemoveGroupMapping(chip::FabricIndex fabric_index, const GroupMapping & mapping)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(kInvalidEndpointId != mapping.endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    FabricData fabric(fabric_index);
+
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_INVALID_FABRIC_ID);
```

Comment:
There may be a way to improve scoping here: we only need `fabric.first_endpoint`, not the whole `FabricData fabric` on the stack for the duration of the method. If there was a nested scope for just loading the FabricData and the `first_endpoint` was outside that scope, then after loading that value, stack for the FabricData object would be released, so the total stack usage would be less. This is true for some of the other `PersistentData` instance in here, where we extract a single member value, but keep the entire class on  the stack for the whole method.

Reply:
Since I have moved the buffers to the Load, and Save methods, the data structures are now cheap (with the exception of the KeySets), so I think this optimization is no longer necessary.